cmd/restic/cmd_copy.go

```diff
@@ -243,10 +223,110 @@ func copyTree(ctx context.Context, srcRepo restic.Repository, dstRepo restic.Rep
 
 	bar := printer.NewCounter("packs copied")
 	bar.SetMax(uint64(len(packList)))
-	_, err = repository.Repack(ctx, srcRepo, dstRepo, packList, copyBlobs, bar, printer.P)
+	_, err = repository.RepackInner(ctx, srcRepo, dstRepo, packList, copyBlobs, bar, printer.P)
 	bar.Done()
 	if err != nil {
 		return errors.Fatalf("%s", err)
 	}
 	return nil
 }
+
+// collectAllSnapshots: select all snapshot trees to be copied in one go
+func collectAllSnapshots(ctx context.Context, opts CopyOptions,
+	srcSnapshotLister restic.Lister, srcRepo restic.Repository,
+	dstSnapshotByOriginal map[restic.ID][]*data.Snapshot, args []string, printer progress.Printer) (selectedSnapshots []*data.Snapshot) {
+
+	selectedSnapshots = make([]*data.Snapshot, 0, 10)
+	for sn := range FindFilteredSnapshots(ctx, srcSnapshotLister, srcRepo, &opts.SnapshotFilter, args, printer) {
+		// check whether the destination has a snapshot with the same persistent ID which has similar snapshot fields
+		srcOriginal := *sn.ID()
+		if sn.Original != nil {
+			srcOriginal = *sn.Original
+		}
+
+		if originalSns, ok := dstSnapshotByOriginal[srcOriginal]; ok {
+			isCopy := false
+			for _, originalSn := range originalSns {
+				if similarSnapshots(originalSn, sn) {
+					printer.V("\n%v\n", sn)
+					printer.V("skipping source snapshot %s, was already copied to snapshot %s\n", sn.ID().Str(), originalSn.ID().Str())
+					isCopy = true
+					break
+				}
+			}
+			if isCopy {
+				continue
+			}
+		}
+		selectedSnapshots = append(selectedSnapshots, sn)
+	}
+
+	return selectedSnapshots
+}
+
+// copyTreeBatched: copy multiple snapshot trees in one go, using calls to
+// repository.RepackInner() for all selected snapshot trees and thereby packing the packfiles optimally.
+// Usually each snapshot creates at least one tree packfile and one data packfile.
+func copyTreeBatched(ctx context.Context, srcRepo restic.Repository, dstRepo restic.Repository,
+	visitedTrees restic.IDSet, selectedSnapshots []*data.Snapshot, opts CopyOptions,
+	printer progress.Printer) error {
+
+	// dependent on opts.batch the package Uploader is started either for
+	// each snapshot to be copied or once for all snapshots
+	var wg *errgroup.Group
+	var wgCtx context.Context
+	var seenBlobs restic.IDSet
+
+	if opts.batch {
```

Comment:
With `repository.RepackInner` (or it's replacement) we can actually let copy automatically do the right thing. Now, that `copyTree` no longer flushes automatically, `copyTreeBatched` can just check that a reasonable amount of data has already been copied before flushing the repository. This should only add a minimal amount of extra code but has the benefit of offering optimal performance out of the box. 

In case you're wondering how to implement this flexible flush control with `WithBlobUploader`, take a look at the double loop at https://github.com/restic/restic/blob/c854338ad152098928c49136a2a2149db8243224/internal/repository/repack_test.go#L21 .